src/main.cpp

```diff
@@ -4619,54 +4600,72 @@ bool SendMessages(CNode* pto, bool fSendTrickle)
         //
         // Message: inventory
         //
-        vector<CInv> vInv;
-        vector<CInv> vInvWait;
-        {
-            LOCK(pto->cs_inventory);
-            vInv.reserve(pto->vInventoryToSend.size());
-            vInvWait.reserve(pto->vInventoryToSend.size());
-            BOOST_FOREACH(const CInv& inv, pto->vInventoryToSend)
-            {
-                if (pto->setInventoryKnown.count(inv))
-                    continue;
+        vector<CInv> vInv, vInvWait;
+        nNowMicros = GetTimeMicros();
 
-                // trickle out tx inv to protect privacy
-                if (inv.type == MSG_TX && !fSendTrickle)
-                {
-                    // 1/4 of tx invs blast to all immediately
-                    static uint256 hashSalt;
-                    if (hashSalt.IsNull())
-                        hashSalt = GetRandHash();
-                    uint256 hashRand = ArithToUint256(UintToArith256(inv.hash) ^ UintToArith256(hashSalt));
-                    hashRand = Hash(BEGIN(hashRand), END(hashRand));
-                    bool fTrickleWait = ((UintToArith256(hashRand) & 3) != 0);
+        vInv.reserve(1000);
+        
+        if (pto->nNextInvSend < nNowMicros)
+            vInvWait.reserve(pto->vInventoryToSend.size());
 
-                    if (fTrickleWait)
+        {
+            LOCK(pto->cs_inventory);
+            
+            // TODO it would be sufficient to walk vInventoryToSend randomly, shuffling is relatively inefficient
```

Comment:
In a random walk, it is hard to efficiently avoid processing the same element twice. If shuffling is slow, create an array of indexes, and shuffle that. Also, where does std::random_shuffle() get its randomness?

Suggested code:
```
std::vector<int> vRandInv;
vRandInv.resize(pto->vInventoryToSend.size());
for (int i = 0; i < pto->vRandInv.size(); i++) vRandInv[i] = i;
for (int i = 0; i < pto->vRandInv.size(); i++) std::swap(vRandInv[i], vRandInv[i + GetRand(vRandInv.size() - i]);

for (int i = 0; i < pto->vRandInv.size(); i++) {
    const CInv& inv = pto->vInventoryToSend[vRandInv[i]];

    ...
```